server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -479,74 +483,97 @@ public void cloneShardSnapshot(
+                    if (existingTargetFiles.isSame(sourceFiles)) {
+                        return existingShardGen;
+                    }
+                    throw new RepositoryException(
+                        metadata.name(),
+                        "Can't create clone of ["
+                            + shardId
+                            + "] for snapshot ["
+                            + target
+                            + "]. A snapshot by that name already exists for this shard."
+                    );
+                }
+                final BlobStoreIndexShardSnapshot sourceMeta = loadShardSnapshot(shardContainer, source);
+                logger.trace("[{}] [{}] writing shard snapshot file for clone", shardId, target);
+                INDEX_SHARD_SNAPSHOT_FORMAT.write(
+                    sourceMeta.asClone(target.getName(), startTime, threadPool.absoluteTimeInMillis() - startTime),
+                    shardContainer,
+                    target.getUUID(),
+                    compress
+                );
+                INDEX_SHARD_SNAPSHOTS_FORMAT.write(
+                    existingSnapshots.withClone(source.getName(), target.getName()),
+                    shardContainer,
+                    newGen,
+                    compress
                 );
-                existingShardGen = String.valueOf(tuple.v2());
-                newGen = String.valueOf(tuple.v2() + 1);
-                existingSnapshots = tuple.v1();
             } else {
                 newGen = UUIDs.randomBase64UUID();
-                existingSnapshots = buildBlobStoreIndexShardSnapshots(Collections.emptySet(), shardContainer, shardGeneration).v1();
-                existingShardGen = shardGeneration;
-            }
-            SnapshotFiles existingTargetFiles = null;
-            SnapshotFiles sourceFiles = null;
-            for (SnapshotFiles existingSnapshot : existingSnapshots) {
-                final String snapshotName = existingSnapshot.snapshot();
-                if (snapshotName.equals(target.getName())) {
-                    existingTargetFiles = existingSnapshot;
-                } else if (snapshotName.equals(source.getName())) {
-                    sourceFiles = existingSnapshot;
-                }
-                if (sourceFiles != null && existingTargetFiles != null) {
-                    break;
-                }
-            }
-            if (sourceFiles == null) {
-                throw new RepositoryException(
-                    metadata.name(),
-                    "Can't create clone of ["
-                        + shardId
-                        + "] for snapshot ["
-                        + target
-                        + "]. The source snapshot ["
-                        + source
-                        + "] was not found in the shard metadata."
+                RemoteStoreShardShallowCopySnapshot remStoreBasedShardMetadata = loadRemStoreEnabledShardSnapshot(shardContainer, source);
```

Comment:
We will need to have following checks for clone now: 
1. source snapshot (full copy) and a target snapshot with same name but shallow snapshot already exists.
2. source snapshot (shallow copy) and a target snapshot with same name but full copy already exists.
3. source snapshot (shallow copy) and a target snapshot with same name but shallow copy already exists.
4. source snapshot (full copy) and a target snapshot with same name but full snapshot already exists. (This is only covered currently)